lib/guardian/plug/pipeline.ex

```diff
@@ -0,0 +1,241 @@
+
+    # Now change out the error handler for plugs downstream of this one.
+    plug Guardian.Plug.Pipeline, error_handler: MyApp.SpecialAuthErrorHandler
+    ```
+    """
+
+    import Plug.Conn
+
+    alias Guardian.Plug, as: GPlug
+
+    @doc """
+    Create your very own `Guardian.Plug.Pipeline`
+
+    Using this macro will make your module into a plug builder.
+    It will provide your pipeline with the Guardian implementation module and error
+    handler so that it can be used within your pipeline and downstream.
+    """
+    defmacro __using__(opts \\ []) do
+      alias Guardian.Plug.Pipeline
+
+      quote do
+        use Plug.Builder
+        alias Guardian.Config, as: GConfig
+        alias Guardian.Plug, as: GPlug
+
+        import Pipeline
+
+        plug :put_modules
+
+        def init(options) do
+          new_opts =
+            options
+            |> Keyword.merge(unquote(opts))
+            |> config()
+
+          unless Keyword.get(new_opts, :otp_app),
+            do: raise_error(:otp_app)
+
+          new_opts
+        end
+
+        defp config(opts) do
+          case Keyword.get(opts, :otp_app) do
+            nil -> opts
+            otp_app ->
+              otp_app
+              |> Application.get_env(__MODULE__, [])
+              |> Keyword.merge(opts)
+          end
+        end
+
+        defp config(opts, key, default \\ nil) do
+          unquote(opts)
+          |> Keyword.merge(opts)
+          |> config()
+          |> Keyword.get(key)
+          |> GConfig.resolve_value()
+        end
+
+        defp put_modules(conn, opts) do
+          pipeline_opts = [
+            module: config(opts, :module),
+            error_handler: config(opts, :error_handler),
+            key: config(opts, :key, GPlug.default_key()),
+          ]
+
+          Pipeline.call(conn, pipeline_opts)
+        end
+
+        defp raise_error(key),
+          do: raise "Config `#{key}` is missing for #{__MODULE__}"
+      end
+    end
+
+    def init(opts), do: opts
+
+    def call(conn, opts) do
+      conn
+      |> maybe_put_key(:guardian_module, Keyword.get(opts, :module))
+      |> maybe_put_key(:guardian_error_handler, Keyword.get(opts, :error_handler))
+      |> maybe_put_key(:guardian_key, Keyword.get(opts, :key))
+    end
+
+    def put_key(conn, key),
+      do: put_private(conn, :guardian_key, key)
+
+    def put_module(conn, module),
+      do: put_private(conn, :guardian_module, module)
+
+    def put_error_handler(conn, module),
+      do: put_private(conn, :guardian_error_handler, module)
+
+    def current_key(conn), do: conn.private[:guardian_key]
+
+    def current_module(conn), do: conn.private[:guardian_module]
+
+    def current_error_handler(conn), do: conn.private[:guardian_error_handler]
+
+    def fetch_key(conn, opts),
+      do: Keyword.get(opts, :key) || current_key(conn) || GPlug.default_key()
+
+    def fetch_module(conn, opts),
+      do: Keyword.get(opts, :module) || current_module(conn)
```

Comment:
You could always forgo the `||` and just use the current mod as the default param:

```elixir
Keyword.get(opts, :module, current_module(conn))
```